pkg/wasm/imagefetcher.go

```diff
@@ -22,20 +22,90 @@ import (
 	"crypto/tls"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"path/filepath"
 	"reflect"
 	"strings"
+	"time"
 
 	"github.com/docker/cli/cli/config/configfile"
 	dtypes "github.com/docker/cli/cli/config/types"
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/google/go-containerregistry/pkg/name"
 	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/google"
 	"github.com/google/go-containerregistry/pkg/v1/remote"
 	"github.com/google/go-containerregistry/pkg/v1/types"
 	"github.com/hashicorp/go-multierror"
+
+	ecr "github.com/awslabs/amazon-ecr-credential-helper/ecr-login"
+	acr "github.com/chrismellard/docker-credential-acr-env/pkg/credhelper"
+)
+
+const (
+	// According to https://docs.aws.amazon.com/cli/latest/reference/ecr/get-authorization-token.html#output,
+	// the expiration interval of the auth token in Amazon is 12 hour. For safety, let's uses 6 hours.
+	ecrCredExpiration = time.Hour * 6
+	// According to https://docs.microsoft.com/en-us/azure/active-directory/develop/refresh-tokens#refresh-token-lifetime,
+	// the expiration interval of the auth token in Azure is 24 hour. For the safety, let's uses 12 hours.
+	// Note that Azure uses the Azure AD refresh token as a password when authenticating a docker client.
+	acrCredExpiration = time.Hour * 12
+)
+
+// Make a default key chain with the support for GCR, ECR, and ACR.
+// After looking up DOCKER_CONFIG by `authn.DefaultKeychain`,
+// GCR, ECR, and ACR are tried sequentially.
+// ECR and ACR helpers does not provide simple way to cache the credential before expiration at this moment.
+// So, `cachedHelper` keeps the credential for the specified duration.
+// In case of google.Keychain, by ReuseTokenSource, the credential is cached.
+// Refer to https://github.com/google/go-containerregistry/blob/4d7b65b04609719eb0f23afa8669ba4b47178571/pkg/v1/google/auth.go#L60.
+var defaultKeychain = authn.NewMultiKeychain(
+	authn.DefaultKeychain,
+	google.Keychain,
+	authn.NewKeychainFromHelper(wrapHelperWithCache(ecr.NewECRHelper(ecr.WithLogger(ioutil.Discard)), ecrCredExpiration)),
```

Comment:
Do you know if NewECRHelper by default retries? If not, we will have to setup ClientFactory that enables retries.

Reply:
Good point.

According to the implementation (https://github.com/aws/aws-sdk-go-v2/blob/cdc9a4c1860acf6ff6ab559b801c5134e515a349/service/ecr/api_client.go#L188), we can see that there is default retryer.https://github.com/aws/aws-sdk-go-v2/blob/cdc9a4c1860acf6ff6ab559b801c5134e515a349/aws/retry/standard.go#L125

Reply:
Similarly, in case of ACR, there is retry logic.

1. creating a container registry specific client: https://github.com/aws/aws-sdk-go-v2/blob/cdc9a4c1860acf6ff6ab559b801c5134e515a349/aws/retry/standard.go#L125
2. creating an autorest client: https://github.com/Azure/go-autorest/blob/7dd32b67be4e6c9386b9ba7b1c44a51263f05270/autorest/client.go#L211-L222

Reply:
Cool. I assume the same with Azure and GCP

Reply:
Yes, `google.Keychain` also retry when getting failure.